NIFI-11022 Add DecryptContent Compatibility Processors

exceptionfactory opened the pull request:
# Summary

[NIFI-11022](https://issues.apache.org/jira/browse/NIFI-11022) Adds the following new Processors to support decryption of content encrypted using the `EncryptContent` Processor:

- `DecryptContentCompatibilityMode`
- `DecryptContentEncoded`

A new `nifi-cipher-bundle` with `nifi-cipher-nar` module contains the new Processors. A new `nifi-security-crypto-key` module in `nifi-commons` provides standard interfaces and implementations of strong Key Derivation Functions using Bouncy Castle components.

The purpose of the new Processors is to maintain components that are capable of decrypting historical information without needing to maintain encryption of insecure and custom formats.

## DecryptContentCompatibilityMode

The `DecryptContentCompatibilityMode` Processor focuses on legacy Password-Based Encryption algorithms, many of which are defined in [RFC 8018](https://www.rfc-editor.org/rfc/rfc8018). The Processor supports a `Key Derivation Strategy` using either [OpenSSL EVP BytesToKey](https://www.openssl.org/docs/man3.0/man3/EVP_BytesToKey.html) or [Jasypt Java simplified encryption](http://www.jasypt.org/). These settings map to the OpenSSL and NiFi Legacy settings in the `EncryptContent` Processor. Neither of these key derivation strategies meet the security requirements of modern cryptography, so the `DecryptContentCompatibilityMode` Processor exists exclusively to provide compatibility with historical modes of encryption. This Processor relies on the Bouncy Castle Security Provider to support cipher operations, but avoids global registration of the provider.

## DecryptContentEncoded

The `DecryptContentEncoded` Processor focuses on newer Key Derivation Functions, supporting Argon2, bcrypt, PBKDF2, and scrypt algorithms. The Processor supports selection of `Cipher Algorithm Mode` and `Cipher Algorithm Padding`, defaulting to `GCM` and `NoPadding` respectively. These settings match the default configuration of `EncryptContent` for the AES symmetric cipher. The `EncryptContent` Processor and supporting implementation components write encrypted content using a format that is specific to Apache NiFi, and also write the content header using a standard byte format specific to each support Key Derivation Function. The `DecryptContentEncoded` Processor leverages this standard byte formatting to determine the appropriate Key Derivation Function for decryption. The Processor supports a `Key Specification Format` property, defaulting to `PASSWORD`, but also allowing `RAW` to enable configuration of a raw key encoded as a hexadecimal string. Both the `DecryptContentEncoded` Processor and the supporting `nifi-security-crypto-key` library include extensive tests to exercise byte format detection and cipher algorithm configurations.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [X] JDK 11
  - [X] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

exceptionfactory commented:
The [DecryptContent Flow](https://issues.apache.org/jira/secure/attachment/13054291/DecryptContent_Flow.json) attached to [NIFI-11022](https://issues.apache.org/jira/browse/NIFI-11022) includes multiple Process Group configurations with `EncryptContent` and the new `DecryptContent` Processors.

greyp9 commented:
@exceptionfactory this all looks really cool, thanks for the effort!

Did some usage testing, and wanted to share some feedback.  I plan to follow up with a pass at the code.

- The sample flow attached to the JIRA was really helpful to me to give a sense of the depth of permutations.  It occurs to me that those nine sample configurations might form the basis of some helpful user documentation.  Something like a table that describes, for a particular `EncryptContent` configuration, what an equivalent `DecryptContentEncoded`/`DecryptContentCompatibility` configuration should be.

  - EncryptContent 
Key Derivation Function = NiFi Legacy KDF
Encryption Algorithm = MD5_DES

  - DecryptContentCompatibilityMode
Encryption Scheme = PBE_WITH_MD5_AND_DES
Key Derivation Strategy = JASYPT_STANDARD

  If not in the user documentation somewhere, maybe some other reference-able place.  The nine samples give a good sense of the configurability; no need to document every permutation.

- After a FlowFile has gone through an encrypt / decrypt processor pair, should any of the `encryptcontent.*` attributes be stripped out of the FF attribute set?  Not sure if that breaks an existing FF convention, but if those are left over, they might confuse downstream processors, or lead to IV reuse, for example.

- It looks like the encryption algorithms are specified in `EncryptionMethod` (for `EncryptContent`) and `CompatibilityModeEncryptionScheme` (for `DecryptContentCompatibilityMode`).  What do you think of these being alphabetized in the UI listbox?  The choices matching `DecryptContentCompatibilityMode` `PBE_WITH_MD5_*` appear in two separate groupings in the UI listbox.

- `org.jasypt` has released a `1.9.3` version.  Might this be a good time to consider bumping the version dependency?

- When specifying a raw key (for `Key Specification Format`), it might be nice for the tooltip to provide a hint as to the number of hex characters expected.  There might not be a single answer, given that many algorithms (AES) have implementations with multiple key sizes.  Maybe some blurb along the lines of:
  > Advanced Encryption Standard (AES) keys are symmetric keys that can be three different key lengths (128, 192, or 256 bits).

  This may be better handled as follow on work.

- In my usage testing, I was swapping back and forth between use of `Password` versus `Raw Key (hexidecimal)`, using parameters `#{MyPassword}` and `#{MyRawKey}`.  These are two distinct properties in `EncryptContent`.  But in `DecryptContentEncoded`, one property is overridden to provide storage for both `Key Specification Format` options.  It might be nice for these to be two distinct properties, so changing the format does not require changing the key specification value.  These could be presented in the UI the same way using `PropertyDependency`.

exceptionfactory commented:
Thanks for the thorough testing and feedback @greyp9!

In response to your notes:

1. I agree that including some additional documentation for these Processors describing the mapping between configuration settings would be helpful. Creating additional details documents for these two Processors seems like a good starting location. There is more general encryption documentation in the Administrator's Guide, which could be updated separately when we are ready to deprecate the EncryptContent Processor.

2. The EncryptContent FlowFile attributes were at one time intended to support simplified integration, but that was never completed implemented. The Initialization Vector cannot be configured in the DecryptContent Processors, and both expect to read it from the FlowFile content. For these reasons, I don't think it is necessary to remove any of the attributes that EncryptContent applies.

3. The EncryptionMethod values have retained their current order for a long time, so I think those should remain unchanged. However, I agree that the values for DecryptContentCompatibilityMode should be alphabetized, so I will make those changes.

4. The jasypt dependency is not actually used, so I created [NIFI-11090](https://issues.apache.org/jira/browse/NIFI-11090) to remove the dependency.

5. With the DecryptContent Processors, any raw hexadecimal key must match the value used for encryption. Adding a note might be helpful as a follow on effort, but I'm not sure it is particularly helpful given that it should already be known, otherwise decryption will fail.

6. I considered having separate properties versus sharing the same property and interpreting it based on the `Key Specification Format` property. I would not expect the value and format to change in normal usage, so it seems more straightforward to keep the current approach, as opposed to introducing a separate property. If you think it would be significantly better to have a separate property, and have both properties depend on the format property, I can look at making those changes.

exceptionfactory commented:
Thanks again for the feedback @greyp9, I pushed an update with additional details documentation for both Processors, including tables that map new property values to matching values from `EncryptContent`.

greyp9 commented on nifi-nar-bundles/nifi-cipher-bundle/nifi-cipher-processors/src/main/java/org/apache/nifi/processors/cipher/compatibility/CompatibilityModeEncryptionScheme.java:
```diff
@@ -0,0 +1,216 @@
+            "PBEWITHMD5ANDDES",
+            DigestAlgorithm.MD5,
+            DES,
+            "PKCS5 Scheme 1 with MD5 digest and Data Encryption Standard 64 bit keys. OID 1.2.840.113549.1.5.3"
```
> PKCS5

Is this a typo?

exceptionfactory replied:
No, `PKCS5` is correct. Some schemes follow the PKCS5 strategy for password handling, others follow PKCS12.

greyp9 commented on nifi-nar-bundles/nifi-cipher-bundle/nifi-cipher-processors/src/main/resources/docs/org.apache.nifi.processors.cipher.DecryptContentCompatibilityMode/additionalDetails.html:
```diff
@@ -0,0 +1,144 @@
+    <h2>Summary</h2>
+    <p>
+      This Processor supports decryption using legacy formats and Password-Based Encryption Algorithms.
+      <a href="https://www.rfc-editor.org/rfc/rfc8018">RFC 8018</a> defines the Password-Based Cryptography
```
Weird interaction here between iframe and external link.

exceptionfactory replied:
Good point, adding the `target` attribute to the tag should help by opening it in a new window.

greyp9 commented on nifi-nar-bundles/nifi-cipher-bundle/nifi-cipher-processors/src/main/java/org/apache/nifi/processors/cipher/DecryptContentEncoded.java:
```diff
@@ -0,0 +1,364 @@
+
+    private static class GCMSerializedParameterSpec extends GCMParameterSpec implements SerializedParameterSpec {
+
+        private static final int GCM_TAG_LENGTH_BITS = 128;
```
https://crypto.stackexchange.com/questions/41601/aes-gcm-recommended-iv-size-why-12-bytes

exceptionfactory commented:
Thanks for the feedback @greyp9, I updated the RFC 8018 link and renamed the Processors as follows:

- Renamed `DecryptContentEncoded` to `DecryptContent`
- Renamed `DecryptContentCompatibilityMode` to `DecryptContentCompatibility`

greyp9 reviewed (approved):
Thanks for those updates.  This all looks good.  I think I like the updated processor names better.
